### ISSUE TITLE
Connectivity test stuck at Waiting for NodePort since 1.15.5

### ISSUE THREAD

**Issue by miminar** [cilium/cilium#33184]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

After cilium chart update to 1.15.5 or 1.15.6, the connectivity test reaches timeout waiting for NodePort.

### Cilium Version

1.15.6

### Kernel Version

`Linux eu-k8s-dev-01.euler.ethz.ch 5.15.0-112-generic #122-Ubuntu SMP Thu May 23 07:48:21 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux`

### Kubernetes Version

`v1.30.1+k0s`

### Regression

Yes, it's a regression, we see this issue neither with 1.15.1 nor with 1.15.4 

### Relevant log output

```shell
$ cilium connectivity test -n cilium
ℹ️   Monitor aggregation detected, will skip some flow validation steps
ℹ️   Skipping tests that require a node Without Cilium
⌛ [eu-k8s-dev] Waiting for deployment cilium-test/client to become ready...
⌛ [eu-k8s-dev] Waiting for deployment cilium-test/client2 to become ready...
⌛ [eu-k8s-dev] Waiting for deployment cilium-test/echo-same-node to become ready...
⌛ [eu-k8s-dev] Waiting for deployment cilium-test/client3 to become ready...
⌛ [eu-k8s-dev] Waiting for deployment cilium-test/echo-other-node to become ready...
⌛ [eu-k8s-dev] Waiting for pod cilium-test/client-856666cb56-d6mtk to reach DNS server on cilium-test/echo-same-node-9975f89cd-xvb49 pod...
⌛ [eu-k8s-dev] Waiting for pod cilium-test/client2-b6f5f8c45-frgls to reach DNS server on cilium-test/echo-same-node-9975f89cd-xvb49 pod...
⌛ [eu-k8s-dev] Waiting for pod cilium-test/client3-7c4bfdfdf4-r8r4n to reach DNS server on cilium-test/echo-same-node-9975f89cd-xvb49 pod...
⌛ [eu-k8s-dev] Waiting for pod cilium-test/client3-7c4bfdfdf4-r8r4n to reach DNS server on cilium-test/echo-other-node-65965cdd64-6jbhj pod...
⌛ [eu-k8s-dev] Waiting for pod cilium-test/client-856666cb56-d6mtk to reach DNS server on cilium-test/echo-other-node-65965cdd64-6jbhj pod...
⌛ [eu-k8s-dev] Waiting for pod cilium-test/client2-b6f5f8c45-frgls to reach DNS server on cilium-test/echo-other-node-65965cdd64-6jbhj pod...
⌛ [eu-k8s-dev] Waiting for pod cilium-test/client-856666cb56-d6mtk to reach default/kubernetes service...
⌛ [eu-k8s-dev] Waiting for pod cilium-test/client2-b6f5f8c45-frgls to reach default/kubernetes service...
⌛ [eu-k8s-dev] Waiting for pod cilium-test/client3-7c4bfdfdf4-r8r4n to reach default/kubernetes service...
⌛ [eu-k8s-dev] Waiting for Service cilium-test/echo-other-node to become ready...
⌛ [eu-k8s-dev] Waiting for Service cilium-test/echo-other-node to be synchronized by Cilium pod cilium/cilium-ctscw
⌛ [eu-k8s-dev] Waiting for Service cilium-test/echo-other-node to be synchronized by Cilium pod cilium/cilium-v8jhs
⌛ [eu-k8s-dev] Waiting for Service cilium-test/echo-same-node to become ready...
⌛ [eu-k8s-dev] Waiting for Service cilium-test/echo-same-node to be synchronized by Cilium pod cilium/cilium-ctscw
⌛ [eu-k8s-dev] Waiting for Service cilium-test/echo-same-node to be synchronized by Cilium pod cilium/cilium-v8jhs
⌛ [eu-k8s-dev] Waiting for NodePort 10.205.160.27:32722 (cilium-test/echo-other-node) to become ready...
⌛ [eu-k8s-dev] Waiting for NodePort 10.205.160.27:30349 (cilium-test/echo-same-node) to become ready...
⌛ [eu-k8s-dev] Waiting for NodePort 10.205.160.20:32722 (cilium-test/echo-other-node) to become ready...
timeout reached waiting for NodePort 10.205.160.20:32722 (cilium-test/echo-other-node) (last error: error with exec request (pod=cilium-test/client-856666cb56-d6mtk, container=): context deadline exceeded)


$ hubble observe  --verdict 'DROPPED'  -f
Jun 14 14:14:52.299: cilium-test/client-856666cb56-d6mtk:38931 (ID:1162528) <> 10.205.160.20:32722 (host) policy-verdict:none INGRESS DENIED (TCP Flags: SYN)
Jun 14 14:14:52.299: cilium-test/client-856666cb56-d6mtk:38931 (world) <> 10.205.160.20:32722 (host) Policy denied DROPPED (TCP Flags: SYN)
Jun 14 14:14:52.693: 9c:b6:54:fb:a2:79 (host) <> 01:80:c2:00:00:00 (world) Unsupported L2 protocol DROPPED (Ethernet)
...
Jun 14 14:14:52.820: 00:26:f0:11:00:00 (host) <> 01:80:c2:00:00:00 (world) Unsupported L2 protocol DROPPED (Ethernet)
Jun 14 14:14:53.319: cilium-test/client-856666cb56-d6mtk:38931 (ID:1162528) <> 10.205.160.20:32722 (host) policy-verdict:none INGRESS DENIED (TCP Flags: SYN)
Jun 14 14:14:53.319: cilium-test/client-856666cb56-d6mtk:38931 (world) <> 10.205.160.20:32722 (host) Policy denied DROPPED (TCP Flags: SYN)
Jun 14 14:14:53.693: 9c:b6:54:fb:a2:77 (host) <> 01:80:c2:00:00:00 (world) Unsupported L2 protocol DROPPED (Ethernet)
...
Jun 14 14:14:55.693: 9c:b6:54:fb:a2:73 (host) <> 01:80:c2:00:00:00 (world) Unsupported L2 protocol DROPPED (Ethernet)
Jun 14 14:14:56.464: cilium-test/client-856666cb56-d6mtk:44977 (ID:1162528) <> 10.205.160.20:32722 (host) policy-verdict:none INGRESS DENIED (TCP Flags: SYN)
Jun 14 14:14:56.464: cilium-test/client-856666cb56-d6mtk:44977 (world) <> 10.205.160.20:32722 (host) Policy denied DROPPED (TCP Flags: SYN)
Jun 14 14:14:56.693: 9c:b6:54:fb:a2:79 (host) <> 01:80:c2:00:00:00 (world) Unsupported L2 protocol DROPPED (Ethernet)
...
Jun 14 14:14:57.058: 02:a0:98:ef:5b:e6 (unknown) <> ff:ff:ff:ff:ff:ff (unknown) Unsupported L3 protocol DROPPED (Ethernet)
Jun 14 14:14:57.479: cilium-test/client-856666cb56-d6mtk:44977 (ID:1162528) <> 10.205.160.20:32722 (host) policy-verdict:none INGRESS DENIED (TCP Flags: SYN)
Jun 14 14:14:57.479: cilium-test/client-856666cb56-d6mtk:44977 (world) <> 10.205.160.20:32722 (host) Policy denied DROPPED (TCP Flags: SYN)
Jun 14 14:14:57.693: 9c:b6:54:fb:a2:77 (host) <> 01:80:c2:00:00:00 (world) Unsupported L2 protocol DROPPED (Ethernet)
...
Jun 14 14:15:00.693: 9c:b6:54:fb:a2:72 (host) <> 01:80:c2:00:00:00 (world) Unsupported L2 protocol DROPPED (Ethernet)
Jun 14 14:15:00.764: cilium-test/client-856666cb56-d6mtk:40913 (ID:1162528) <> 10.205.160.20:32722 (host) policy-verdict:none INGRESS DENIED (TCP Flags: SYN)
Jun 14 14:15:00.764: cilium-test/client-856666cb56-d6mtk:40913 (world) <> 10.205.160.20:32722 (host) Policy denied DROPPED (TCP Flags: SYN)
```

### Anything else?

When trying to upload sysdump here, I get : `File size too big: 25 MB are allowed, 35 MB were attempted to upload.`

Hence uploaded it [here](https://drive.google.com/file/d/1IkdlEUwVKSWM4aqB7equCYfuH68lCdRt/view?usp=sharing)

### Cilium Users Document

- [X] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ti-mo** [comment]
@miminar Thanks for the report. Is this just the connectivity tests breaking, or NodePort breaking as a whole?

**Comment by miminar** [comment]
> @miminar Thanks for the report. Is this just the connectivity tests breaking, or NodePort breaking as a whole?

It's not just the connectivity test. Although a connection to a node port always works from the host network, a connection from the pod network fails. And only from a particular node (the first one):

```
root@net-debug:/# for h in eu-k8s-dev-{01..10}.euler.ethz.ch; do curl -I --connect-timeout 1 http://$h:32346 |& grep '^HTTP' || echo "$h: BAD"; done
eu-k8s-dev-01.euler.ethz.ch: BAD
HTTP/1.1 200 OK
HTTP/1.1 200 OK
HTTP/1.1 200 OK
HTTP/1.1 200 OK
HTTP/1.1 200 OK
HTTP/1.1 200 OK
HTTP/1.1 200 OK
HTTP/1.1 200 OK
HTTP/1.1 200 OK
```

Tested with services of type NodePort and LoadBalancer - same results. FWIW, the node has been re-added to the cluster:

```sh
$ kubectl get nodes
NAME                          STATUS   ROLES                   AGE     VERSION
eu-k8s-dev-01.euler.ethz.ch   Ready    compute,control-plane   2d20h   v1.30.1+k0s
eu-k8s-dev-02.euler.ethz.ch   Ready    compute,control-plane   17d     v1.30.1+k0s
eu-k8s-dev-03.euler.ethz.ch   Ready    compute,control-plane   17d     v1.30.1+k0s
eu-k8s-dev-04.euler.ethz.ch   Ready    compute                 2d20h   v1.30.1+k0s
eu-k8s-dev-05.euler.ethz.ch   Ready    compute                 16d     v1.30.1+k0s
eu-k8s-dev-06.euler.ethz.ch   Ready    compute                 16d     v1.30.1+k0s
eu-k8s-dev-07.euler.ethz.ch   Ready    compute                 16d     v1.30.1+k0s
eu-k8s-dev-08.euler.ethz.ch   Ready    compute                 16d     v1.30.1+k0s
eu-k8s-dev-09.euler.ethz.ch   Ready    compute                 9d      v1.30.1+k0s
eu-k8s-dev-10.euler.ethz.ch   Ready    compute                 9d      v1.30.1+k0s
```

But the same goes for `eu-k8s-dev-04.euler.ethz.ch` where this cannot be reproduced.

The connectivity test happens to choose the first node for the node port readiness check, that's why it fails reliably.

After a rollback to 1.15.4, I cannot reproduce anymore with any kind of node port.

**Comment by miminar** [comment]
Same issue on 1.15.7

**Comment by miminar** [comment]
It looks to be related to a k8s VIP, which is managed by kube-vip in our case.
The first node has the VIP (10.205.160.17/32):

```
root@eu-k8s-dev-01:~# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: access: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc mq state UP group default qlen 1000
    link/ether 94:57:a5:d3:4b:ee brd ff:ff:ff:ff:ff:ff
    altname enp1s0
    altname eno1
    inet 10.205.160.20/20 metric 100 brd 10.205.175.255 scope global access
       valid_lft forever preferred_lft forever
    inet 10.205.160.17/32 scope global access
       valid_lft forever preferred_lft forever
```

And the connectivity to node ports works against the VIP, not against the static host IP 10.205.160.20:

```
[adm-mminar@eu-ops-01 bin]$ for i in 10.205.160.17 10.205.160.{20..27} 10.205.160.{39..40}; do echo -n "$i: "; curl -sI --connect-timeout 1 http://$i:32346 | grep ^HTTP || echo FAILED; done
10.205.160.17: HTTP/1.1 200 OK
10.205.160.20: FAILED
10.205.160.21: HTTP/1.1 200 OK
10.205.160.22: HTTP/1.1 200 OK
10.205.160.23: HTTP/1.1 200 OK
10.205.160.24: HTTP/1.1 200 OK
10.205.160.25: HTTP/1.1 200 OK
10.205.160.26: HTTP/1.1 200 OK
10.205.160.27: HTTP/1.1 200 OK
10.205.160.39: HTTP/1.1 200 OK
10.205.160.40: HTTP/1.1 200 OK
```

Where :32 belongs to redmine NodePort service:

```
kubectl get svc -o wide -A | grep NodePort | grep redmine
redmine                redmine-np                                           NodePort       10.205.209.129   <none>          3000:32346/TCP                                                                                                  50d     app.kubernetes.io/component=redmine,app.kubernetes.io/instance=redmine,app.kubernetes.io/name=redmine
```

**Comment by miminar** [comment]
Resolved in 1.15.8:

````sh
[adm-mminar@eu-ops-01 ~]$ for i in 10.205.160.17 10.205.160.{20..27} 10.205.160.{39..40}; do echo -n "$i: "; curl -sI --connect-timeout 1 http://$i:32346 | grep ^HTTP || echo FAILED; done
10.205.160.17: FAILED
10.205.160.20: HTTP/1.1 200 OK
10.205.160.21: HTTP/1.1 200 OK
10.205.160.22: HTTP/1.1 200 OK
10.205.160.23: HTTP/1.1 200 OK
10.205.160.24: HTTP/1.1 200 OK
10.205.160.25: HTTP/1.1 200 OK
10.205.160.26: HTTP/1.1 200 OK
10.205.160.27: HTTP/1.1 200 OK
10.205.160.39: HTTP/1.1 200 OK
10.205.160.40: HTTP/1.1 200 OK
```

Where 10.205.160.17 is the VIP.
Same result after triggering a fail-over to another node.
